x-pack/plugins/security_solution/server/endpoint/services/artifacts/unified_manifest_client.ts

```diff
@@ -0,0 +1,197 @@
+import { mapUnifiedManifestSavedObjectToUnifiedManifest } from './utils';
+import type {
+  InternalUnifiedManifestCreateSchema,
+  InternalUnifiedManifestBaseSchema,
+  InternalUnifiedManifestSchema,
+  InternalUnifiedManifestUpdateSchema,
+} from '../../schemas';
+import { ManifestConstants } from '../../lib/artifacts';
+
+interface FetchAllUnifiedManifestsOptions {
+  perPage?: number;
+  kuery?: string;
+  sortField?: string;
+  sortOrder?: 'desc' | 'asc';
+  fields?: string[];
+}
+
+export const UNIFIED_MANIFEST_ALL_NAMESPACES = '*';
+
+export class UnifiedManifestClient {
+  private savedObjectsClient: SavedObjectsClientContract;
+
+  constructor(savedObjectsClient: SavedObjectsClientContract) {
+    this.savedObjectsClient = savedObjectsClient;
+  }
+
+  /**
+   * Create
+   */
+
+  public createUnifiedManifest(
+    manifest: InternalUnifiedManifestBaseSchema
+  ): Promise<SavedObjectsBulkResponse<InternalUnifiedManifestCreateSchema>> {
+    return this.createUnifiedManifests([manifest]);
+  }
+
+  public createUnifiedManifests(
+    manifest: InternalUnifiedManifestBaseSchema[]
+  ): Promise<SavedObjectsBulkResponse<InternalUnifiedManifestCreateSchema>> {
+    return this.savedObjectsClient.bulkCreate<InternalUnifiedManifestCreateSchema>(
+      manifest.map((attributes) => ({
+        type: ManifestConstants.UNIFIED_SAVED_OBJECT_TYPE,
+        attributes: {
+          ...attributes,
+          created: Date.now(),
+        },
+      })),
+      { initialNamespaces: [UNIFIED_MANIFEST_ALL_NAMESPACES] }
+    );
+  }
+
+  /**
+   * Read
+   */
+
+  public getUnifiedManifestByPolicyId(
+    policyId: string
+  ): Promise<SavedObjectsFindResponse<InternalUnifiedManifestSchema>> {
+    return this.savedObjectsClient.find({
+      type: ManifestConstants.UNIFIED_SAVED_OBJECT_TYPE,
+      search: policyId,
+      searchFields: ['policyId'],
+      namespaces: [UNIFIED_MANIFEST_ALL_NAMESPACES],
+    });
+  }
+
+  public getUnifiedManifestById(
+    manifestId: string
+  ): Promise<SavedObjectsBulkResponse<InternalUnifiedManifestSchema>> {
+    return this.getUnifiedManifestByIds([manifestId]);
+  }
+
+  public getUnifiedManifestByIds(
+    manifestIds: string[]
+  ): Promise<SavedObjectsBulkResponse<InternalUnifiedManifestSchema>> {
+    return this.savedObjectsClient.bulkGet(
+      manifestIds.map((id) => ({ id, type: ManifestConstants.UNIFIED_SAVED_OBJECT_TYPE })),
+      { namespace: UNIFIED_MANIFEST_ALL_NAMESPACES }
+    );
+  }
+
+  public async getAllUnifiedManifests(
+    cb: (unifiedManifests: InternalUnifiedManifestSchema[]) => void | Promise<void>,
+    options?: FetchAllUnifiedManifestsOptions
+  ): Promise<void> {
+    const unifiedManifestsFetcher = this.fetchAllUnifiedManifests(this.savedObjectsClient, options);
+
+    for await (const unifiedManifests of unifiedManifestsFetcher) {
+      if (cb.constructor.name === 'AsyncFunction') {
+        await cb(unifiedManifests);
+      } else {
+        cb(unifiedManifests);
+      }
```

Comment:
awaiting a non promise func would result in the func being wrapped in a Promise which might lead to non expected execution order. Not likely to happen but still, this check makes sure we control execution stack. 
Let me know if my knowledge is outdated though :)

Reply:
yeah, I needed to re-read some [docs](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/await) :) I think there should be no issue with using `await` before a normal function: it will be wrapped in a Promise as you said, but as the function passed to the Promise does not contain async parts, it should be immediately performed in the same tick, so messing up the execution order could not happen.

I think we don't win with this check, so it should be safe to remove it and win some readability - but I'm not insisting on this if keeping it makes you more confident in this code :)